src/System.Private.Xml/src/System/Xml/XmlEncoding.cs

```diff
@@ -494,16 +495,16 @@ public override void Convert(byte[] bytes, int byteIndex, int byteCount, char[]
                     return;
                 }
                 // decode 1 character from the byte cache
-                i = GetFullChars(lastBytes, 0, 4, chars, charIndex);
-                Debug.Assert(i == 1);
+                i = GetFullChars(lastBytes, 0, 4, chars, charIndex, out surrogateCount);
+                Debug.Assert(i == 1 + surrogateCount);
                 charIndex += i;
                 charCount -= i;
                 charsUsed = i;
 
                 lastBytesCount = 0;
 
                 // if that's all that was requested -> return
-                if (charCount == 0)
+                if (charCount + surrogateCount == 0)
```

Comment:
It seems like the entirety of the fix could have been encapsulated by a one-line change:

```cs
if (charCount <= 0)
```

Is it really expected that `charCount` could ever be negative?

Reply:
That would work for this scenario, but as I don't know all possible cases so added this new variable just to avoid from broking anything else.

> Is it really expected that charCount could ever be negative?

I assume the person who wrote it wasn't expecting that, but it is becaming negative in case of surrogate chars

Reply:
I will update with your suggestion and run tests

Reply:
Right - I agree with your assessment that I don't think this number was intended to go negative. But with the proposed fix we still allow this to happen. If this behavior is indeed intentional, then a check for negative values should be sufficient to fix the bug. If the number's not intended to go negative, then it seems we need to tweak the fix further so that this condition doesn't occur.

Reply:
Seems main intention of this condition was only expected number of chars converted successfully. But the creator of this code didn't count surrogate chars case, so we might better to check if (i == 1 || i == 2) instead of  if (charCount <= 0)

Reply:
Spoke with @buyaa-n offline. This type subclasses `Decoder`, but it doesn't properly fulfill the contract of `Decoder.Convert` and related methods. For example, it sets `completed = true` in some scenarios even when there's still unconsumed state in the decoder. The existing implementation of this type requires a much deeper investigation than this small handful of changes.

Reply:
Yes it is written in very inefficient way, but it is working and have no major issue reported so far. The bug related to this PR only happening for error nous condition (bytes with invalid char), so not really sure if its need/worth for deeper investigation ...